library/public/loader.h

```diff
@@ -12,76 +13,64 @@ namespace f3d
- * Then files from the list can be loaded one by one using loadFile method.
- * Information about the files in the list can be recovered using getFileInfo.
- * The current file can be changed either on loadFile with a dedicated enum or
- * by using setCurrentFileIndex.
+ * A class to load files in F3D. The loader can load a full scene or multiple geometries into a
+ * default scene. It also support checking if a scene or geometry reader is available for a given
+ * file.
+ *
+ * Example usage:
+ *
+ *  std::string path = ...
+ *  f3d::engine eng(f3d::window::Type::NATIVE);
+ *  f3d::loader& load = eng.getLoader();
+ *
+ *  if (load.hasSceneReader(path)
+ *  {
+ *    load.loadFullScene(path);
+ *  }
+ *  else if (load.hasFullSceneReader(path)
+ *  {
+ *    load.loadGeometry(path);
+ *  }
+ *
  */
 class F3D_EXPORT loader
 {
 public:
   /**
-   * An enum used in loadFile to select which file to load.
+   * An exception that can be thrown by the loader
+   * when it failed to load a file for some reason.
    */
-  enum class LoadFileEnum
+  struct load_failure_exception : public exception
   {
-    LOAD_FIRST,
-    LOAD_PREVIOUS,
-    LOAD_CURRENT,
-    LOAD_NEXT,
-    LOAD_LAST
+    load_failure_exception(const std::string& what = "")
+      : exception(what){};
   };
 
   /**
-   * Add a list of files or directories to be loaded.
-   * Set recursive to true to load folder recursively instead of only files.
-   * Duplicate will not be added.
-   */
-  virtual loader& addFiles(const std::vector<std::string>& files, bool recursive = false) = 0;
-
-  /**
-   * Add a file or directory (all files in) to be loaded.
-   * Set recursive to true to load folder recursively instead of only files.
-   */
-  virtual loader& addFile(const std::string& path, bool recursive = false) = 0;
-
-  /**
-   * Get the vector of files to be loaded.
-   */
-  virtual const std::vector<std::string>& getFiles() const = 0;
-
-  /**
-   * Set the current file index, corresponding to the file that will be loaded
-   * when calling loadFile(LOAD_CURRENT).
+   * Load a geometry from a provided file to the scene.
+   * Reset the scene before loading if a full scene was loaded previously or if reset is set to
+   * false, do not reset if only loaded geometries previously. Geometries loader using this method
+   * will be available in a default scene and use all default scene related options. Throw a
+   * load_failure_exception on failure.
    */
-  virtual loader& setCurrentFileIndex(int index) = 0;
+  virtual loader& loadGeometry(const std::string& filePath, bool reset = false) = 0;
 
   /**
-   * Get the current file index.
+   * Reset scene and load a full scene from provided file.
+   * Please note default scene related options are not taken into account when loading a full scene.
+   * Throw a load_failure_exception on failure.
    */
-  virtual int getCurrentFileIndex() const = 0;
+  virtual loader& loadFullScene(const std::string& filePath) = 0;
```

Comment:
We have `hasGeometryReader` and `loadGeometry`, but `hasSceneReader` and `loadFullScene`.  
Let's rename the API `loadScene`, it's more consistent and shorter.

Reply:
I'd like to have the same `has[...]Reader` and `load[...]` pattern. In the first case you have `Geometry` and it's fine. For the second case you have `hasSceneReader` an `loadFullScene` so `Scene`/`FullScene` which is inconsistent.  
It should be either `hasSceneReader`/`loadScene` or `hasFullSceneReader`/`loadFullScene` (I prefer the first proposition)

Reply:
got it, I agree.